cli/src/pcluster/templates/pcluster_dashboard_builder.py

```diff
@@ -0,0 +1,586 @@
+        """Add the textwidget to the cloudwatch dashboard and update coordinates."""
+        text_widget = TextWidget(markdown=markdown, height=1, width=24)
+        text_widget.position(x=self.coord["x"], y=self.coord["y"])
+        cloudwatch_dashboard.add_widgets(text_widget)
+        self._update_coord_after_section(self.coord, 1)
+
+    def _generate_graph_widget(self, title, metric_list):
+        graph_width = 6
+        graph_height = 6
+        widget = GraphWidget(
+            title=title,
+            left=metric_list,
+            region=self._stack_region,
+            width=graph_width,
+            height=graph_width,
+        )
+        widget.position(x=self.coord["x"], y=self.coord["y"])
+        self._update_coord(self.coord, graph_width, graph_height)
+        return widget
+
+    def _generate_ec2_metrics_list(self, metrics):
+        metric_list = []
+        for metric in metrics:
+            cloudwatch_metric = cloudwatch.Metric(
+                namespace="AWS/EC2", metric_name=metric, dimensions={"InstanceId": self.head_node_instance.ref}
+            )
+            metric_list.append(cloudwatch_metric)
+        return metric_list
+
+    def _add_conditional_storage_widgets(
+        self,
+        conditional_metrics,
+        volumes_list,
+        namespace,
+        dimension_vol_name,
+        attribute_name,
+        vol_attribute_name,
+    ):
+        """Add widgets for conditional metrics for EBS, Raid and EFS."""
+        widgets_list = []
+        for metric_condition_params in conditional_metrics:
+
+            # Check if there are volumes support the metric
+            is_supported_vol_present = False
+            for volume in volumes_list:
+                if not is_supported_vol_present:
+                    if getattr(volume, vol_attribute_name) in metric_condition_params["supported_vol_types"]:
+                        is_supported_vol_present = True
+
+            # If supported volume for this metric present, add graph widget
+            if is_supported_vol_present:
+                metric_list = []
+                for volume in volumes_list:
+                    if getattr(volume, vol_attribute_name) in metric_condition_params["supported_vol_types"]:
+                        cloudwatch_metric = cloudwatch.Metric(
+                            namespace=namespace,
+                            metric_name=metric_condition_params["metric"],
+                            dimensions={dimension_vol_name: getattr(volume, attribute_name, "")},  # TODO:FIXME
+                        )
+                        metric_list.append(cloudwatch_metric)
+                graph_widget = self._generate_graph_widget(metric_condition_params["title"], metric_list)
+                widgets_list.append(graph_widget)
+        return widgets_list
+
+    def _add_storage_widgets(self, metrics, storages_list, namespace, dimension_name, attribute_name):
+        widgets_list = []
+        for metrics_param in metrics:
```

Comment:
nit: you have different ways to loop into a dictionary: https://realpython.com/iterate-through-dictionary-python/#iterating-through-items